### PR TITLE
fix(metadata): align CalRevision with Java (MD5 + toDescString format)

### LINKED ISSUE #3577: [Bug] Metadata revision mismatch between dubbo-go v3.3 and Java Dubbo 3.3 (SHA-512 vs MD5, divergent toDescString)
## 背景
dubbo-go 是 Apache Dubbo 的 Go 实现。Consumer 以 `revision`（元数据修订 hash）对 Provider 元数据进行去重缓存：相同 revision 的实例共用一份元数据，避免重复调用 MetadataService。缓存键见 `registry/servicediscovery/service_instances_changed_listener_impl.go:266` 的 `metadataCacheKey(app, registryId, revision)`。

## 问题描述
Go 与 Java Provider 暴露“相同”服务时，各自算出的 revision 不同（长度、内容均异）。Consumer 按 revision 分键缓存，同一应用被缓存为两份，丧失去重收益，跨语言互通场景元数据拉取次数翻倍。

## 根因分析（v3.3 源码比对，3 处不一致）
1. **Hash 算法**：Go 用 SHA-512（128 hex），Java 用 MD5（32 hex 小写）。
   - Java：`RevisionResolver.calRevision` → `md5Utils.getMd5(metadata)`
   - Go：`CalRevision(...)` → SHA-512（`metadata/info/metadata_info.go`）
2. **toDescString 格式**：Go 用 `|` 分隔并独立追加 methods；Java 用 `matchKey+port+path+无分隔符 params`。
   - Java：`getMatchKey() + port + path + new TreeMap<>(getParams())`
   - Go：`"name|group|version|protocol|port|path|params|methods"`
3. **methods 是否参与序列化**：Go 参与；Java 不参与（params 仅含 `method.key` 参数，不含 methods 列表）。
   - Go：`NewServiceInfoWithURL` 将 MethodsKey 写入 params，且 `toDescString` 单独拼接 methods。

注：`matchKey` 格式 `serviceKey:protocol` 两边一致，无需改。

## 影响
跨语言互通场景元数据拉取次数翻倍、注册中心/MetadataService 压力上升。v3 应用级服务发现已大幅缓解早期应用注册时机差异（v1.5 时代），但 revision 不一致仍在。

## 复现步骤
1. 启动 Java 与 dubbo-go Provider 暴露同接口同配置的 `com.foo.Bar`；
2. 分别读取两端 `calRevision` / `CalRevision` 输出；
3. 比对：长度（128 vs 32 hex）与内容均不同 → 不一致。

## 建议修复方案
- `toDescString()` 改为 Java 格式：`matchKey + port + path + sortedParams`（排除 methods、无分隔符、不单独拼接）。
- `CalRevision` 由 SHA-512 改为 MD5（UTF-8 字节 MD5，32 hex 小写）。
- `NewServiceInfoWithURL` 使 params 与 Java 一致（methods 不入序列化；`GetMethods` 保留读取途径）。
- 新增跨语言 golden vector 测试：固定一组 service，断言 Go `CalRevision` 输出 == 已知 Java MD5 常量（取自 Java 单测），作为回归守护。

## 验收标准
- [ ] Go 与 Java 对相同服务的 revision 一致（32 hex、值相同）。
- [ ] 跨语言 golden vector 测试通过。
- [ ] 纯 Go 集群在 develop 分支完成一次性元数据重新拉取验证，无功能回退。
- [ ] 兼容性说明写入 changelog / PR 描述。

## 兼容性说明
改动会使所有 dubbo-go 应用 revision 变化，纯 Go 集群需一次性重新拉取元数据（短暂），但与 Java 互通为硬约束，应在 develop 分支（未正式发布）尽快修正。

### PR DESCRIPTION
**Pull request by AlexStocks**
## What problem does this PR solve?

This aligns dubbo-go v3.3's metadata `revision` computation with Java Dubbo 3.3 so the two sides produce **byte-for-byte identical** revision strings.

Previously dubbo-go computed `revision` with **SHA-512** and a pipe-separated `toDescString`, while Java used **MD5** with a different string format. The mismatched revision caused a Consumer to cache the same service's metadata **twice** (one Go copy + one Java copy) under `metadataCacheKey(app, registryId, revision)`. Interop still worked, but metadata efficiency was lost.

Fixes #3577

## Root cause & fix (3 inconsistencies resolved)

1. **Hash algorithm**: SHA-512 → MD5 (32-char lowercase hex), matching `RevisionResolver.calRevision()` in Java.
2. **`toDescString` serialization**: changed from `name|group|version|protocol|port|path|params|methods` to the Java form `getMatchKey() + port + path + TreeMap(params).toString()` — i.e. `matchKey + strconv.Itoa(port) + path + "{k=v, k=v}"` (no separators, methods excluded). `matchKey = serviceKey + ":" + protocol` was already identical to Java and is unchanged.
3. **`methods` participation**: Java excludes the method list from the revision string. Added a `ServiceInfo.Methods []string` field (hessian/json tagged `methods`); `NewServiceInfoWithURL` stops writing `MethodsKey` into params and sets `si.Methods` instead. `GetMethods()` / `DeepCopy()` updated; `GetParams()` no longer injects methods.

## Verification

- Golden vector derived **independently from the Java spec** (not from Go output): `app + sorted-by-matchKey toDescString(...)` → MD5 `8ab6351c86cccf642be0a2cda8be847a`, matching Java.
- Tests: `go test ./metadata/info/... ./registry/servicediscovery/...` → **all pass** (73 test cases). Includes hessian serialization of `ServiceInfo`, `toDescString` format locks, Java-alignment golden test, empty-services → `"0"`, ordering independence, and method-list exclusion.
- Multi-service ordering matches Java: the `services` map is keyed by `matchKey` and sorted by it (same as Java's TreeMap).

## Out of scope (known residual difference)

- The `protocol`-empty `:` edge case in `common.MatchKey` is **not** changed this round — service-discovery URLs always carry a protocol, and the change has too-wide impact. Documented as a follow-up.

## Test plan

- [x] `go test ./metadata/info/...`
- [x] `go test ./registry/servicediscovery/...`

### CODE REVIEW

**Comment by Alanxtl** [comment]
@Qiao-yq help review this

File: metadata/info/metadata_info.go [hunk 1]
```diff
@@ -287,8 +288,8 @@ func NewServiceInfoWithURL(url *common.URL) *ServiceInfo {
 			}
 		}
 	}
-	p[constant.MethodsKey] = strings.Join(url.Methods, ",")
 	service.Params = p
+	service.Methods = url.Methods
```
Comment (Qiao-yq) [inline metadata/info/metadata_info.go:292]: image
不应将 methods 从 Params 迁移到独立字段，
Java Dubbo 3.3 的 ServiceInfo 没有独立的 methods 字段，方法列表保存在 params 中，并参与 toDescString() 和 revision 计算。另外，MetadataServiceV2 的 protobuf 没有 methods 字段，当前转换逻辑也只传输 Params。迁移后，方法列表会在 V2 metadata 往返过程中丢失。

File: metadata/info/metadata_info.go [hunk 2]
```diff
@@ -364,90 +364,79 @@ func (si *ServiceInfo) DeepCopy() *ServiceInfo {
...
-	keys := make([]string, 0, len(si.Params))
-	for k := range si.Params {
-		if k == constant.MethodsKey {
-			continue
-		}
+// renderParams renders a param map as a Java TreeMap.toString() equivalent:
+// "{k1=v1, k2=v2}" with keys in natural ascending order, joined by ", ".
+// An empty map renders as "{}". This mirrors Java's
+// `new TreeMap<>(params).toString()` used by ServiceInfo.toDescString(), which
+// is what makes the revision digest byte-for-byte identical to the Java side.
```
Comment (Qiao-yq) [inline metadata/info/metadata_info.go:399]: https://github.com/apache/dubbo/blob/3.3/dubbo-metadata/dubbo-metadata-api/src/main/java/org/apache/dubbo/metadata/DefaultMetadataParamsFilter.java#L38-L65
这里虽然对齐了参数序列化格式，但现有 Go IncludeKeys 与 Java 默认过滤规则有一些不同，Go 包含 timestamp，Java 默认排除它。

**Comment by Qiao-yq** [comment]
以上是我的一点理解，如有遗漏或理解不对的地方，还请指正ovo
